File: Prog3/TrabalhoJAVA/src/app/Eleicao.java
package app;

import java.util.ArrayList;
import java.util.List;

/**
 * A classe Eleição representa um evento de seleção, é responsável por gerar os
 * relatórios e tem todas as informações necessárias para isso
 */
public class Eleicao {
    private int qtdEleitos;
    private List<Candidato> candidatos;
    private List<Partido> partidos;
    private int qtdCandidados;
    private int qtdPartidos;

    Eleicao(List<Candidato> candidatos, List<Partido> partidos) {
        this.candidatos = candidatos;
        this.partidos = partidos;
        this.qtdCandidados = candidatos.size();
        this.qtdPartidos = partidos.size();
    }

    /**
     * Essa função serve para procurar um partido dentro de uma lista de partidos
     *
     * @param partidos   A lista de partidos onde deseja-se realizar a busca
     * @param numPartido O número do partido que está sendo buscado
     * @return Caso o partido seja encontrado dentro da lista, ele é retornado, caso
     *         contrário, é retornado um ponteiro NULL
     */
    public static Partido retornaPartidoLista(List<Partido> partidos, int numPartido) {
        Partido partido = null;

        for (Partido p : partidos) {
            if (p.getNumPartido() == numPartido) {
                partido = p;
                break;
            }
        }
        return partido;
    }

    /**
     * Essa função serve para encontrar os candidados eleitos daquela eleição, além
     * disso ela atualiza o atributo de qtdEleitos de Eleição
     * 
     * @return Retorna uma lista dos candidados eleitos
     */
    public List<Candidato> retornaCandidatosEleitos() {
        List<Candidato> candidatosEleitos = new ArrayList<Candidato>();
        int cont = 0;

        for (Candidato c : this.candidatos) {
            if (c.getSituacao().compareTo("Eleito") == 0) {
                candidatosEleitos.add(c);
                cont++;
            }
        }

        this.qtdEleitos = cont;

        return candidatosEleitos;
    }

    /**
     * Essa função serve para encontrar os N (N = qtdEleitos), candidatos mais
     * votados
     * 
     * @return Retorna uma lista dos candidados mais votados
     */
    public List<Candidato> retornaCandidatosMaisVotados() {
        List<Candidato> candidatosMaisVotados = new ArrayList<Candidato>();
        int cont = 0;

        for (Candidato c : this.candidatos) {
            if (cont < this.qtdEleitos) {
                candidatosMaisVotados.add(c);
                cont++;
            } else if (cont == this.qtdEleitos)
                break;
        }

        return candidatosMaisVotados;
    }

    /**
     * Essa função serve para encontrar os candidatos que seriam eleitos caso o
     * sistema de votação levasse em conta apenas os votos nominais
     * 
     * @param candidatosMaisVotados Lista dos candidados mais votados
     * @return Retorna uma lista com os candidados que seriam eleitos caso o sistema
     *         levasse em conta apenas os votos nominais
     */
    public List<Candidato> retornaCandidatosSeriamEleitos(List<Candidato> candidatosMaisVotados) {
        List<Candidato> candidatosSeriamEleitos = new ArrayList<Candidato>();

        for (Candidato cmv : candidatosMaisVotados) {
            if (cmv.getSituacao().compareTo("Eleito") == 0)
                candidatosSeriamEleitos.add(null); // Coloca-se NULL, para que os cantidatos preservem suas colocações
            else
                candidatosSeriamEleitos.add(cmv);
        }

        return candidatosSeriamEleitos;
    }

    /**
     * Essa função serve para encontrar os candidatos que foram eleitos se
     * beneficiando do sistema proporcional
     * 
     * @param candidatosMaisVotados Lista dos candidados mais votados
     * @return Retorna uma lista com os candidados que não estão na lista de mais
     *         votados, mas foram eleitos graças ao sistema proporcional
     */
    public List<Candidato> retornaCandidatosBeneficiados(List<Candidato> candidatosMaisVotados) {
        List<Candidato> candidatosBeneficiados = new ArrayList<Candidato>();
        boolean encontrou;

        for (Candidato c : candidatos) {
            encontrou = false;
            if (c.getSituacao().compareTo("Eleito") == 0) {
                for (Candidato cmv : candidatosMaisVotados) {
                    if (c.getNumero() == cmv.getNumero()) {
                        encontrou = true;
                        break;
                    }
                }
                if (encontrou)
                    candidatosBeneficiados.add(null);// Coloca-se NULL, para que os cantidatos preservem suas colocações
                else
                    candidatosBeneficiados.add(c);
            } else
                candidatosBeneficiados.add(null);// Coloca-se NULL, para que os cantidatos preservem suas colocações
        }

        return candidatosBeneficiados;
    }

    /**
     * Essa função serve para calcular as quantidades de candidatos eleitos por
     * faixa etária
     * 
     * @param candidatosEleitos Lista dos candidados eleitos
     * @return retorna um vetor de inteiros com a quantidade de candidatos eleitos
     *         em cada intervalo de faixa etária, ou seja, a posição 0 são
     *         candidados entre 30 e 40 anos, a posição 1 de 40 a 50 e assim por
     *         diante
     * 
     */
    public int[] retornaQtdCandidatosEleitosFaixaEtaria(List<Candidato> candidatosEleitos) {
        int[] qtdCandidatosFaixaEtaria = new int[5];
        int qtd;
        boolean limiteIdade;

        for (int i = 30, j = 0, k = 0; k < 5; i += 10, j = i - 10, k++) {
            qtd = 0;
            for (Candidato ce : candidatosEleitos) {
                limiteIdade = (i > 60) ? true : (ce.getIdade() < i);
                if (ce.getIdade() >= j && limiteIdade) {
                    qtd++;
                }
            }
            qtdCandidatosFaixaEtaria[k] = qtd;
        }
        return qtdCandidatosFaixaEtaria;
    }

    /**
     * Essa função serve para calcular as quantidades de candidatos eleitos por sexo
     * 
     * @param candidatosEleitos Lista dos candidados eleitos
     * @return retorna um vetor de inteiros com a quantidade de candidatos eleitos
     *         do sexo masculino e feminino, sendo a posição 0 do vetor o sexo
     *         feminino e a posição 1 o sexo masculino
     * 
     */
    public int[] retornaQtdCandidatosSexo(List<Candidato> candidatosEleitos) {
        int mas = 0;
        int fem = 0;

        int[] qtdPorSexo = new int[2];

        for (Candidato c : candidatosEleitos) {
            if (c.getSexo() == 'F')
                fem++;
            if (c.getSexo() == 'M')
                mas++;
        }
        qtdPorSexo[0] = fem;
        qtdPorSexo[1] = mas;

        return qtdPorSexo;
    }

    public int getQtdEleitos() {
        return this.qtdEleitos;
    }

    public List<Candidato> getCandidatos() {
        return candidatos;
    }

    public List<Partido> getPartidos() {
        return partidos;
    }

    public int getQtdCandidados() {
        return qtdCandidados;
    }

    public int getQtdPartidos() {
        return qtdPartidos;
    }
}

File: Prog3/TrabalhoC++/eleicao.cpp
#include "eleicao.h"

Eleicao::Eleicao(list<Candidato*> candidatos, list<Partido> partidos) {
    this->candidatos = candidatos;
    this->partidos = partidos;
    this->qtd_candidados = candidatos.size();
    this->qtd_partidos = partidos.size();
}

//Tentamos passar partidos como const, porém dava erro de compilação e não soubemos resolver
Partido Eleicao::retorna_partido_lista(list<Partido>& partidos, int num_partido) {
    Partido partido;

    for (Partido& p : partidos) {
        if (p.get_num_partido() == num_partido) {
            return p;

            break;
        }
    }
    return partido;
}

list<Candidato*> Eleicao::retorna_candidatos_eleitos() {
    list<Candidato*> candidatos_eleitos;
    int cont = 0;

    for (Candidato* c : this->candidatos) {
        if (c->get_situacao() == "Eleito") {
            candidatos_eleitos.push_back(c);
            cont++;
        }
    }

    this->qtd_eleitos = cont;

    return candidatos_eleitos;
}

list<Candidato*> Eleicao::retorna_candidatos_mais_votados() {
    list<Candidato*> candidatos_mais_votados;
    int cont = 0;

    for (Candidato* c : this->candidatos) {
        if (cont < this->qtd_eleitos) {
            candidatos_mais_votados.push_back(c);
            cont++;
        } else if (cont == this->qtd_eleitos)
            break;
    }

    return candidatos_mais_votados;
}

list<Candidato*> Eleicao::retorna_candidatos_seriam_eleitos(const list<Candidato*>& candidatos_mais_votados) {
    list<Candidato*> candidatos_seriam_eleitos;

    for (Candidato* cmv : candidatos_mais_votados) {
        if (cmv->get_situacao() == "Eleito")
            candidatos_seriam_eleitos.push_back(NULL);  // Coloca-se NULL, para que os candidatos preservem suas colocações
        else
            candidatos_seriam_eleitos.push_back(cmv);
    }

    return candidatos_seriam_eleitos;
}

list<Candidato*> Eleicao::retorna_candidatos_beneficiados(const list<Candidato*>& candidatos_mais_votados) {
    list<Candidato*> candidatos_beneficiados;
    bool encontrou;

    for (Candidato* c : this->candidatos) {
        encontrou = false;
        if (c->get_situacao() == "Eleito") {
            for (Candidato* cmv : candidatos_mais_votados) {
                if (c->get_numero() == cmv->get_numero()) {
                    encontrou = true;
                    break;
                }
            }
            if (encontrou)
                candidatos_beneficiados.push_back(NULL);  // Coloca-se NULL, para que os candidatos preservem suas colocações
            else
                candidatos_beneficiados.push_back(c);
        } else
            candidatos_beneficiados.push_back(NULL);  // Coloca-se NULL, para que os candidatos preservem suas colocações
    }

    return candidatos_beneficiados;
}

int* Eleicao::retorna_qtd_candidatos_eleitos_faixa_etaria(const list<Candidato*>& candidatos_eleitos) {
    int* qtd_candidatos_faixa_etaria = (int*)malloc(sizeof(int) * 5);
    int qtd;
    bool limite_idade;

    for (int i = 30, j = 0, k = 0; k < 5; i += 10, j = i - 10, k++) {
        qtd = 0;
        for (Candidato* ce : candidatos_eleitos) {
            limite_idade = (i > 60) ? true : (ce->get_idade() < i);
            if (ce->get_idade() >= j && limite_idade) {
                qtd++;
            }
        }
        qtd_candidatos_faixa_etaria[k] = qtd;
    }
    return qtd_candidatos_faixa_etaria;
}

//erroooooo
int* Eleicao::retorna_qtd_candidatos_sexo(const list<Candidato*>& candidatos_eleitos) {
    int mas = 0;
    int fem = 0;

    int* qtd_por_sexo = (int*)malloc(sizeof(int) * 2);

    for (Candidato* c : candidatos_eleitos) {
        if (c->get_sexo() == 'F')
            fem++;
        if (c->get_sexo() == 'M')
            mas++;
    }
    qtd_por_sexo[0] = fem;
    qtd_por_sexo[1] = mas;

    return qtd_por_sexo;
}

int Eleicao::get_qtd_eleitos() const {
    return this->qtd_eleitos;
}

list<Candidato*> Eleicao::get_candidatos() const {
    return this->candidatos;
}

list<Partido> Eleicao::get_partidos() const {
    return this->partidos;
}

int Eleicao::get_qtd_candidados() const {
    return this->qtd_candidados;
}

int Eleicao::get_qtd_partidos() const {
    return this->qtd_candidados;
}

Eleicao::~Eleicao() {
}